rust/arrow/src/array/equal/structure.rs

```diff
@@ -30,19 +32,51 @@ fn equal_values(
         .iter()
         .zip(rhs.child_data())
         .all(|(lhs_values, rhs_values)| {
-            equal_range(lhs_values, rhs_values, lhs_start, rhs_start, len)
+            // merge the null data
+            let lhs_merged_nulls = match (lhs_nulls, lhs_values.null_buffer()) {
+                (None, None) => None,
+                (None, Some(c)) => Some(c.clone()),
+                (Some(p), None) => Some(p.clone()),
+                (Some(p), Some(c)) => {
+                    let merged = (p & c).unwrap();
+                    Some(merged)
+                }
+            };
+            let rhs_merged_nulls = match (rhs_nulls, rhs_values.null_buffer()) {
+                (None, None) => None,
+                (None, Some(c)) => Some(c.clone()),
+                (Some(p), None) => Some(p.clone()),
+                (Some(p), Some(c)) => {
+                    let merged = (p & c).unwrap();
+                    Some(merged)
+                }
+            };
+            equal_range(
+                lhs_values,
+                rhs_values,
+                lhs_merged_nulls.as_ref(),
+                rhs_merged_nulls.as_ref(),
+                lhs_start,
+                rhs_start,
+                len,
+            )
         })
 }
 
 pub(super) fn struct_equal(
     lhs: &ArrayData,
     rhs: &ArrayData,
+    lhs_nulls: Option<&Buffer>,
+    rhs_nulls: Option<&Buffer>,
     lhs_start: usize,
     rhs_start: usize,
     len: usize,
 ) -> bool {
-    if lhs.null_count() == 0 && rhs.null_count() == 0 {
-        equal_values(lhs, rhs, lhs_start, rhs_start, len)
+    // we have to recalculate null counts from the null bitmaps
+    let lhs_null_count = count_nulls(lhs_nulls, lhs_start, len);
+    let rhs_null_count = count_nulls(rhs_nulls, rhs_start, len);
+    if lhs_null_count == 0 && rhs_null_count == 0 {
+        equal_values(lhs, rhs, lhs_nulls, rhs_nulls, lhs_start, rhs_start, len)
     } else {
         // with nulls, we need to compare item by item whenever it is not null
         (0..len).all(|i| {
```

Comment:
Wouldn't it be easier if we took into account the nulls on this statement here:

```rust
let lhs_is_null = lhs.is_null(lhs_pos);
let rhs_is_null = rhs.is_null(rhs_pos);
```

by replacing it with

```rust
let lhs_is_null = lhs.is_null(lhs_pos) && lhs.nullbitmap.isnull(lhs_pos);
let rhs_is_null = rhs.is_null(rhs_pos) && rhs.nullbitmap.isnull(rhs_pos);
```

(still keeping the re-count in place, to avoid wrongfully optimizing like this PR fixes)

The advantage is that for all types except `struct`, this PR merges two equal bitmaps twice (all types except `struct` now pass `null_buffer()` to `equal_range`. E.g. in `list`:

```
            lhs_values,
            rhs_values,
            lhs_values.null_buffer(),
            rhs_values.null_buffer(),
```

Reply:
An alternate thing to do is to pass in `None` for all types except `struct`

https://github.com/apache/arrow/pull/8739/files#diff-6fca2b5d7c8b47ba17e7c93c2eaa2cf7f6b65392d0573254fee593449b8adf46R43

Reply:
@alamb if we pass `None` where we don't have a struct, wouldn't we be describing a different position, which might be incorrect?
A `None` for the `Option<&Buffer>` indicates that there are no nulls, as arrays are permitted to omit the null buffer/bitmap altogether if there are no nulls. So if we pass `None` for `lhs_nulls`, we shouldn't have to look at `lhs`, as the presumption is that we've already extracted the null buffer, and there wasn't any.

Perhaps I don't understand you

Reply:
@nevi-me  I think I was confused -- I was trying to offer some way to avoid @jorgecarleitao 's concern of double merge concerns but now that I re-read my comments I am not sure that they make any sense. Sorry for the confusion